modules/substrate/src/verifier.rs

```diff
@@ -0,0 +1,472 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::BridgeStorage;
+use bp_substrate::{prove_finality, AuthoritySet, ImportedHeader, ScheduledChange};
+use sp_finality_grandpa::{ConsensusLog, SetId, GRANDPA_ENGINE_ID};
+use sp_runtime::generic::OpaqueDigestItemId;
+use sp_runtime::traits::Header as HeaderT;
+use sp_std::{prelude::Vec, vec};
+
+/// The finality proof used by the pallet.
+///
+/// For a Substrate based chain using Grandpa this will
+/// be an encoded Grandpa Justification.
+pub type FinalityProof<'a> = &'a [u8];
+
+/// Errors which can happen while importing a header.
+#[derive(Debug, PartialEq)]
+pub enum ImportError {
+	/// This header is older than our latest finalized block, thus not useful.
+	OldHeader,
+	/// This header has already been imported by the pallet.
+	HeaderAlreadyExists,
+	/// We're missing a parent for this header.
+	MissingParent,
+	/// We were unable to prove finality for this header.
+	UnfinalizedHeader,
+	/// We failed to verify this header's ancestry.
+	AncestryCheckFailed,
+	/// The header is missing digests related to consensus events.
+	///
+	/// This will typically have to do with missing authority set change signals.
+	MissingConsensusDigest,
+}
+
+/// A trait for verifying whether a header is valid for a particular blockchain.
+pub trait ChainVerifier<S, H> {
+	/// Import a header to the pallet.
+	fn import_header(storage: &mut S, header: &H, finality_proof: Option<FinalityProof>) -> Result<(), ImportError>;
+
+	/// Verify that the given header has been finalized and is part of the canonical chain.
+	///
+	/// Returns a list of headers which got finalized by the given header.
+	fn verify_finality(storage: &mut S, header: &H, proof: FinalityProof) -> Result<Vec<H>, ImportError>;
+}
+
+/// Used to verify imported headers and their finality status.
+#[derive(Debug)]
+pub struct Verifier;
+
+impl<S, H> ChainVerifier<S, H> for Verifier
+where
+	S: BridgeStorage<Header = H>,
+	H: HeaderT,
+{
+	fn import_header(storage: &mut S, header: &H, finality_proof: Option<FinalityProof>) -> Result<(), ImportError> {
+		let highest_finalized = storage.best_finalized_header().expect("TODO");
+		if header.number() < highest_finalized.number() {
+			return Err(ImportError::OldHeader);
+		}
+
+		if storage.header_exists(header.hash()) {
+			return Err(ImportError::HeaderAlreadyExists);
+		}
+
+		let parent_header = storage.get_header_by_hash(*header.parent_hash());
+		if parent_header.is_none() {
+			return Err(ImportError::MissingParent);
+		}
+
+		// A block at this height should come with a justification and signal a new
+		// authority set. We'll want to make sure it is valid
+		//
+		// This defaults to 0, should it maybe be an Option?
+		let scheduled_change_height = storage.scheduled_set_change().height;
+		if *header.number() == scheduled_change_height && finality_proof.is_some() {
```

Comment:
Looks like you do not support forks. Imagine following simple chain:
```
        / -> B2'
B1 -   -> B2
```

Let's say both `B2` and `B2'`are signalling authorities set change. But `B2` signals it in `10` blocks, while `B2'` signals it in `100`blocks. Seems like your storage only supports one signal. But instead imo it is easier to implement linked list on top of headers hierarchy. Something like `prevSignalHeaderHash` and `prevSignalTargetNumber` [here](https://github.com/svyatonik/substrate-bridge-sol/blob/6456d3e016c95cd5e6d5e817c23e9e69e739aa78/substrate-bridge.sol#L73). This way you could check if justification is required without additional storage reads.

Reply:
Also I'm not sure about condition. Probably it's just unfinished code, but please remember that you can't finalize block `C` that is descendant of block `B` if `B` doesn't have justification but requires it (i.e. it enacts new authority set). Seems like here you just may insert `B` without justification. And you don't have API/Call to finalize existing block.

Reply:
Yeah, I know this pallet doesn't support forks at the moment. I figured I'd start with the "happy path" case first, where all the headers imported are from the same chain, and later on I'd build in fork awareness.

> remember that you can't finalize block C that is descendant of block B if B doesn't have justification but requires it

So I think I have that covered the when checking the ancestry proof. When importing C we check that the last block with a justification, B, and the new block, C, are related. The code is [here](https://github.com/paritytech/parity-bridges-common/pull/296/files#diff-16898582e891075aa30c1b1330962d86R119).

Reply:
> Yeah, I know this pallet doesn't support forks at the moment. I figured I'd start with the "happy path" case first, where all the headers imported are from the same chain, and later on I'd build in fork awareness.

Ok then :) (you were asking if "Am I missing any basic checks in Verifier::import_header" : )

> So I think I have that covered the when checking the ancestry proof. When importing C we check that the last block with a justification, B, and the new block, C, are related. The code is here.

I don't see that :/ Looks like you're just building ancestry from `best_finalized` to new header there.

Reply:
> Ok then :) (you were asking if "Am I missing any basic checks in Verifier::import_header" : )

Ah, I didn't realize you were answering that, lol. I thought it was more of a general observation, since I think fork support goes a little deeper than just `Verifier::import_header` (requires changes in pallet storage, pruning, etc).

> I don't see that :/ Looks like you're just building ancestry from best_finalized to new header there.

In the way the pallet is written now, `best_finalized` will always be the best block we know of that was imported with a justification. This means it's fine if we build from `(best_finalized, being_imported]`, assuming that `best_finalized` and `being_imported` are related.

Although I do realize now that I never update the `BestFinalized` storage item, so maybe that's adding to the confusion.